### ISSUE TITLE
MSLE as loss function

### ISSUE THREAD

**Issue by TKassis** [timeseriesAI/tsai#150]
Thank you for all the contributors to TSAI. Such a fantastic framework!

I have a regression problem where the targets span a really wide range [0.001 to 1000]. I've been using the MSELossFlat and L1LossFlat without any issues. However, I care more about the percent error than the mean average error. As a result, I wrote my own loss function for Mean Square Log Error (MSLE) below:

```python
import torch
import torch.nn as nn

class MSLELoss(nn.Module):
    def __init__(self, reduction='mean'):
        super().__init__()
        assert reduction in ['mean', 'sum', 'none'], "You must set reduction to 'mean', 'sum' or 'none'"
        self.reduction = reduction
        self.mse = nn.MSELoss(reduction=reduction)

    def forward(self, input, target):
        loss = self.mse(torch.log(input.view(-1) + 1), torch.log(target + 1))
        return loss
```

I'm a bit confused about how FastAI/TSAI wants the loss function to look like so not sure I have this correct. The reason I'm skeptical is because my training loop is very unstable I end up with NaNs after a few epochs.

![loss function](https://user-images.githubusercontent.com/9168908/124942532-b905e280-dfd9-11eb-887a-b767ed3ed7c6.png)

Any advice?

Thanks,

**Comment by vrodriguezf** [comment]
Actually fastai has that metric implemented and looks almost like that (although they flatten both the input and the target)

https://github.com/fastai/fastai/blob/2af48aa4747c910283f679505bb5ef77a4b8eeec/fastai/metrics.py#L285

**Comment by TKassis** [comment]
Thanks for pointing this out! But this is only a metric correct? Can I use it as a loss function?

**Comment by oguiza** [comment]
Hi @TKassis,

Thanks for your feedback. I really appreciate it!

I think the issue here is that if the input is < -1 the log results in a nan value. And this may happen at any time during training. If you want to apply log, you need to ensure all inputs are > -1, so that the first term is never a nan.

**Comment by TKassis** [comment]
@oguiza that's explains it! I've changed it as follows:

```python
import torch
import torch.nn as nn
class MSLELoss(nn.Module):
    def __init__(self, reduction='mean'):
        super().__init__()
        assert reduction in ['mean', 'sum', 'none'], "You must set reduction to 'mean', 'sum' or 'none'"
        self.reduction = reduction
        self.mse = nn.MSELoss(reduction=reduction)
    def forward(self, input, target):
        if input.view(-1) <= -1:
            loss = self.mse(torch.log(0 + 1), torch.log(target + 1))
        else:
            loss = self.mse(torch.log(input.view(-1) + 1), torch.log(target + 1))
        return loss
```
Hopefully that fixes it.

**Comment by TKassis** [comment]
Ended up with this which seems to work (at least not give any errors and seems to converge).

```python
import torch
import torch.nn as nn

class MSLELoss(nn.Module):
    def __init__(self, reduction='mean'):
        super().__init__()
        assert reduction in ['mean', 'sum', 'none'], "You must set reduction to 'mean', 'sum' or 'none'"
        self.reduction = reduction
        self.mse = nn.MSELoss(reduction=reduction)

    def forward(self, input, target):
        input[input<=-1] = 0
        loss = self.mse(torch.log(input.view(-1) + 1), torch.log(target + 1))
        return loss
```

**Comment by oguiza** [comment]
Hi @TKassis,
One thing you need to bear in mind if you want to apply this loss, is that when you use the trained model for inference, you will need to post-process the predictions in the same way as in the loss. Otherwise the model may generate unwanted negative predictions.

**Comment by TKassis** [comment]
As in add 1 to the prediction and set all negative predictions to 0?

**Comment by oguiza** [comment]
The issue is with this, as you are changing the predictions. 
When you use the model for inference, some predictions might be negative.
